File: Utilities/EC_Adapter/PushSupplier.java
package com.stdc.Util;

import RtecEventData;
import RtecEventDataHelper;
import RtecEventDataHolder;


// PushSupplier.java
//
// ============================================================================
//
//
// = FILENAME
//    PushSupplier.java
//
// = AUTHOR
//    Amit Haldankar (amit.haldankar@stdc.com)
//
// = DESCRIPTION
//    This is a Push Consumer which takes the data field of the
//    event and updates with it a Data Handler.
//
//
// ============================================================================



// The Consumer has to implement the Skeleton Consumer

public class PushSupplier extends RtecEventComm.PushSupplierPOA
{

  public static final int ACE_ES_EVENT_ANY = 0;
  public static final int ACE_ES_EVENT_SHUTDOWN = 1;
  public static final int ACE_ES_EVENT_ACT = 2;
  public static final int ACE_ES_EVENT_NOTIFICATION = 3;
  public static final int ACE_ES_EVENT_TIMEOUT = 4;
  public static final int ACE_ES_EVENT_INTERVAL_TIMEOUT = 5;
  public static final int ACE_ES_EVENT_DEADLINE_TIMEOUT = 6;
  public static final int ACE_ES_GLOBAL_DESIGNATOR = 7;
  public static final int ACE_ES_CONJUNCTION_DESIGNATOR = 8;
  public static final int ACE_ES_DISJUNCTION_DESIGNATOR = 9;
  public static final int ACE_ES_EVENT_UNDEFINED = 16;
  public static final int TOTAL_MESSAGES = 30;

  private CORBA_Handles _handles = null;
  private int _sendEventType = -1;
  private int _myId = -1;
  private String EventName = null;
  private String _ecName = null;
  private int _disconnect_event_type = PushConsumer.SUPPLIER_DISCONNECT;

  private org.omg.CORBA.ORB _orb;
  private org.omg.PortableServer.POA _poa;
  private RtecScheduler.Scheduler _scheduler;

  private int _rt_info;
  private RtecEventChannelAdmin.EventChannel _eventChannel;
  private RtecEventChannelAdmin.SupplierAdmin _supplier_admin;
  private RtecEventChannelAdmin.ProxyPushConsumer _consumer_proxy;
  private RtecEventChannelAdmin.SupplierQOS _qos;
  private boolean _isScheduled = false;
  private boolean _isSubscribed = false;
  private boolean _connectedToProxy = false;

  public PushSupplier ()
  {
  }

  public PushSupplier (CORBA_Handles CORBAHandles,
                       int sendEventType,
                       int myId,
                       String ecName)
  {
      this.init(CORBAHandles, sendEventType, myId, ecName);
  }

  public void init (CORBA_Handles CORBAHandles,
                    int sendEventType,
                    int myId,
                    String ecName)
  {
      this._handles = CORBAHandles;
      _orb = CORBAHandles.get_orb_handle();
      _poa = CORBAHandles.get_root_poa_handle();
      _scheduler = CORBAHandles.get_scheduler_handle();

      this._sendEventType = sendEventType;
      this._myId = myId;
      this.EventName = "Send QMS Message";
      this._ecName = ecName;
      try {

        // Define Real-time information
        System.out.println("Event Name = " + EventName);
        this._rt_info = _scheduler.create (EventName);

        default_schedule();

        // Register Publishing events

        byte payload[] = new byte[1];
        payload[0] = 0;
        RtecEventComm.Event notification_event_ = new RtecEventComm.Event ();
        notification_event_.header =
          new RtecEventComm.EventHeader (_sendEventType,     //type
                                         _myId,               //source
                                         1,                  //ttl
                                         0L,                 //creation_time
                                         0L,                 //ec_recv_time
                                         0L                  //ec_send_time
                                        );

        notification_event_.data =
          new RtecEventData (0, payload, _orb.create_any());

        RtecBase.Dependency_Info dependency_info = 
                     new RtecBase.Dependency_Info(
                         RtecBase.Dependency_Type_t.ONE_WAY_CALL,          //Dependency Type
                         1,                                                //No of calls
                         _rt_info                                          //rt_info
                         );
        RtecEventChannelAdmin.Publication publications_[] = new RtecEventChannelAdmin.Publication[1];
        publications_[0] = new RtecEventChannelAdmin.Publication (notification_event_, dependency_info);


        // Get SupplierQOS
        _qos = new RtecEventChannelAdmin.SupplierQOS (publications_, false);

      } catch (RtecScheduler.DUPLICATE_NAME e) {
          System.err.println ("PushSupplier.init: Duplicate names");
          System.err.println (e);
      } catch (RtecScheduler.INTERNAL e) {
          System.err.println ("PushSupplier.init: internal scheduler error");
          System.err.println (e);
      } catch (RtecScheduler.SYNCHRONIZATION_FAILURE e) {
          System.err.println ("PushSupplier.init: scheduler synchronization failure");
          System.err.println (e);
      } catch(org.omg.CORBA.SystemException e) {
          System.err.println(e);
      }
  }

  private void default_schedule ()
  {
      try {
          System.out.println("Setting Scheduler Parameters");
          _scheduler.set (_rt_info,
                          RtecScheduler.Criticality_t.VERY_LOW_CRITICALITY,
                          0L,
                          0L,
                          0L,
                          2500000,  // period
                          RtecScheduler.Importance_t.VERY_LOW_IMPORTANCE,
                          0L,
                          1,
                          RtecScheduler.Info_Type_t.OPERATION);
          System.out.println("Scheduler Parameters set");
          _isScheduled = true;
      } catch(org.omg.CORBA.SystemException e) {
          System.err.println("SYS_EX Setting schedule");
          System.err.println(e);
      } catch (RtecScheduler.UNKNOWN_TASK e) {
          System.err.println ("PushSupplier.init: Unknown task");
          System.err.println (e);
      } catch (RtecScheduler.SYNCHRONIZATION_FAILURE e) {
          System.err.println ("PushSupplier.init: scheduler synchronization failure");
          System.err.println (e);
      } catch (RtecScheduler.INTERNAL e) {
          System.err.println ("PushSupplier.default_schedule: internal scheduler error");
          System.err.println (e);
      }
  }

  public void connect ()
  {
      try {
          if ( _isScheduled == false )
              default_schedule();

          // Connect as a supplier
          if ( _supplier_admin == null ) {
              System.out.println("Connecting to EC " + _ecName);
              if ( _eventChannel == null )
                  _eventChannel =
                      _handles.get_event_channel(_ecName);

              System.out.println("Getting Supplier Admin");
              _supplier_admin = _eventChannel.for_suppliers ();
              int count = 0;
              while ( _supplier_admin == null ) {
                  if ( count >= CORBA_Handles.QMS_DEFAULT_RETRIES ) {
                      System.err.println ("Unable to get supplier admin object exiting.");
                      System.exit(1);
                  }
                  try { Thread.sleep(CORBA_Handles.QMS_DEFAULT_RETRY_SLEEP_INTERVAL); 
                  } catch (InterruptedException e) {}
                  System.out.println ("retrying getting consumer admin object");
                  _supplier_admin = _eventChannel.for_suppliers ();
                  count++;
              }
              System.out.println("Got Supplier Admin");
          }

          // Obtain a reference to the proxy push consumer

          if ( _consumer_proxy == null ) {
              System.out.println("Getting Proxy Push Consumer");
              _consumer_proxy = _supplier_admin.obtain_push_consumer ();
              int count = 0;
              while ( _consumer_proxy == null ) {
                  if ( count >= CORBA_Handles.QMS_DEFAULT_RETRIES ) {
                      System.err.println ("Unable to get proxy push consumer exiting.");
                      System.exit(1);
                  }
                  try { Thread.sleep(CORBA_Handles.QMS_DEFAULT_RETRY_SLEEP_INTERVAL); 
                  } catch (InterruptedException e) {}
                  System.out.println ("retrying getting proxy push consumer");
                  _consumer_proxy = _supplier_admin.obtain_push_consumer ();
                  count++;
              }
              System.out.println("Got Proxy Push Consumer");
          }

          if ( !_connectedToProxy ) {
              org.omg.CORBA.Object objref = _poa.servant_to_reference (this);

              RtecEventComm.PushSupplier supplier_ref = 
                RtecEventComm.PushSupplierHelper.narrow (objref);

              _consumer_proxy.connect_push_supplier (supplier_ref, _qos);

              _connectedToProxy = true;
          }

          System.out.println ("Registered the supplier successfully.");

      } catch(org.omg.CORBA.SystemException e) {
          System.err.println("SYS_EX Setting schedule");
          System.err.println(e);
      } catch (org.omg.PortableServer.POAPackage.WrongPolicy e) {
          System.err.println ("PushSupplier.connect: org.omg.PortableServer.POAPackage.WrongPolicy");
          System.err.println (e);
      } catch (org.omg.PortableServer.POAPackage.ServantNotActive e) {
          System.err.println ("PushConsumer.connect: org.omg.PortableServer.POAPackage.ServantNotActive");
          System.err.println (e);
//      } catch (RtecEventChannelAdmin.TypeError e) {
//          System.err.println ("PushConsumer.connect: RtecEventChannelAdmin.TypeError");
//          System.err.println (e);
      } catch (RtecEventChannelAdmin.AlreadyConnected e) {
          System.err.println ("Demo_Consumer.open_consumer: RtecEventChannelAdmin.AlreadyConnected");
          System.err.println (e);
      }
  }

  public int send (String mesg)
  {
        RtecEventComm.Event events[] = new RtecEventComm.Event[1];

        events[0] = new RtecEventComm.Event ();
        events[0].header =
          new RtecEventComm.EventHeader (_sendEventType,     //type
                                         _myId,               //source
                                         1,                  //ttl
                                         0L,                 //creation_time
                                         0L,                 //ec_recv_time
                                         0L                  //ec_send_time
                                        );

        byte payload[] = mesg.getBytes();

        events[0].data =
          new RtecEventData (0, payload, _orb.create_any());

        _consumer_proxy.push (events);

        return 0;
  }

  public void disconnect_push_supplier()
  {
      System.out.println ("Demo Supplier: Have to disconnect!");
      send_disconnect_event();
  }

  public void send_disconnect_event()
  {
        RtecEventComm.Event events[] = new RtecEventComm.Event[1];

        events[0] = new RtecEventComm.Event ();
        events[0].header =
          new RtecEventComm.EventHeader (_disconnect_event_type,     //type
                                         _myId,                      //source
                                         1,                         //ttl
                                         0L,                        //creation_time
                                         0L,                        //ec_recv_time
                                         0L                         //ec_send_time
                                        );

        byte payload[] = new byte[1];
        payload[0] = 0;

        events[0].data =
          new RtecEventData (0, payload, _orb.create_any());

        _consumer_proxy.push (events);

        return;
  }

  public void disconnect()
  {
      if ( _consumer_proxy == null )
          return;

      if (this._connectedToProxy){
          _consumer_proxy.disconnect_push_consumer();
          _connectedToProxy = false;
      }

      if ( _consumer_proxy != null )
          _consumer_proxy = null;

      if ( _supplier_admin != null )
          _supplier_admin = null;

      try {
          _poa.deactivate_object (_poa.servant_to_id (this));
      } catch ( org.omg.PortableServer.POAPackage.ServantNotActive e ) {
          System.err.println ("PushSupplier.disconnect: ServantNotActive");
          System.err.println (e);
      } catch (org.omg.PortableServer.POAPackage.WrongPolicy e) {
          System.err.println ("PushSupplier.disconnect: WrongPolicy");
          System.err.println (e);
      } catch (org.omg.PortableServer.POAPackage.ObjectNotActive e) {
          System.err.println ("PushSupplier.disconnect: ObjectNotActive");
          System.err.println (e);
      }
  }

}

File: Utilities/EC_Adapter/Push_Supplier.cpp
//                              -*- Mode: C++ -*-
// Push_Supplier.cpp ---
// Author           : Manoj Srivastava ( srivasta@ember )
// Created On       : Wed Oct 18 16:37:52 2000
// Created On Node  : ember
// Last Modified By : Manoj Srivastava
// Last Modified On : Thu Dec  6 13:31:44 2001
// Last Machine Used: glaurung.green-gryphon.com
// Update Count     : 136
// Status           : Unknown, Use with caution!
// HISTORY          :
// Description      :
// $Id: Push_Supplier.cpp,v 1.18 2002/02/28 21:02:53 srivasta Exp $
//

/*
 * Copyright (C) 2001 System/Technology Development Corporation
 * This file is part of QoS Metrics Services (QMS)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 * 02111-1307 USA
 *
 * You may contact System/Technology Development Corporation at
 *
 *     System/Technology Development Corporation
 *     Suite 500, Center for Innovative Technology,
 *     2214 Rock Hill Road,
 *     Herndon, VA 20170    
 *     (703) 476-0687
 */


/// RCS version string
static const char* version = 
"$Id: Push_Supplier.cpp,v 1.18 2002/02/28 21:02:53 srivasta Exp $";

/** @file Push_Supplier.cpp
 *  @ingroup TAOEC
 *  @brief Implentation of the Push Supplier class
 *
 *  This is the main implementation for the class.
 */

#include "orbsvcs/Time_Utilities.h"
#include "orbsvcs/Event_Utilities.h"

#include "Push_Supplier.h"
#include "Debug.h"

#if defined (_WIN32) /* Only Win32 needs special treatment. */
static const int  QMS_DEFAULT_number_of_retries     = 3;
static const RtecEventComm::EventType SUPPLIER_DISCONNECT   = 20000;
static const long QMS_DEFAULT_retry_sleep_interval = 60;
#endif

/**
 *  @brief  Constructor
 *
 * Sets up internal variables to a sane default state. The object
 * still needs to be initialized to be useful.
 */
STDC::QMS::Push_Supplier::Push_Supplier(void) :
  Is_Scheduled(false), Is_Subscribed(false), ec_name_p(0), 
  _send_event_type(9999),_disconnect_event_type(9999), _my_id(17), 
  EventName(0), connected_to_proxy(false)
{
  QMS_Trace ("STDC::QMS::Push_Supplier::Push_Supplier", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG: New  Supplier object\n");
  ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);

  this->connected_to_proxy = false;
}
/**
 * @overload
 */

STDC::QMS::Push_Supplier::Push_Supplier(Push_Supplier & rhs):
  POA_RtecEventComm::PushSupplier(),
  handles (rhs. handles),
  _supplier_admin (rhs._supplier_admin),
  _consumer_proxy (rhs._consumer_proxy),
  rt_info (rhs. rt_info),
  Is_Scheduled (rhs. Is_Scheduled),
  qos (rhs. qos),
  Is_Subscribed (rhs. Is_Subscribed),
  _send_event_type (rhs._send_event_type),
  _disconnect_event_type (rhs._disconnect_event_type),
  _my_id (rhs._my_id),
  connected_to_proxy (rhs. connected_to_proxy)
{
  QMS_Trace ("STDC::QMS::Push_Supplier::Push_Supplier", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG: New  Supplier object\n");
  this->EventName              = CORBA::string_dup(rhs.EventName);
}

/**
 * @brief The = operator
 *
 * This avoids self assignements, but does not make anny attempt to
 * disconnect before the copy. It is the callers responsibility to
   ensure that the consumer object is disconnected.
 */
const STDC::QMS::Push_Supplier 
STDC::QMS::Push_Supplier::operator=(const Push_Supplier & copy)
{
  QMS_Trace ("QMS_Consumer::operator =", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG: copy object\n");
  if(&copy != this)
   {  
     ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);

     this->handles          	    = copy.handles;
     this->_supplier_admin  	    = copy._supplier_admin;
     this->_consumer_proxy  	    = copy._consumer_proxy;
     this->rt_info          	    = copy.rt_info;
     this->Is_Scheduled     	    = copy.Is_Scheduled;
     this->qos              	    = copy.qos;
     this->Is_Subscribed    	    = copy.Is_Subscribed;
     this->_send_event_type 	    = copy._send_event_type;
     this->_my_id                 = copy._my_id;
     this->EventName              = CORBA::string_dup(copy.EventName);
     this->connected_to_proxy     = copy.connected_to_proxy;
   } /* end of if(&copy != this) */
  return *this;			// enable cascaded assignments
}
/**
 * @brief Destructor
 */

STDC::QMS::Push_Supplier::~Push_Supplier()
{
  CORBA::string_free(this->EventName);
}


/**
 *  @brief Initialization
 *  @param Handles A set of CORBA references and handles required to set up
 *  @param send_event_type The type of event we shall send out
 *  @param  my_id The supplier ID we shall use for that event
 *
 * This is where the real intialization happens. Well, really, most of
 * it has already happened in the caller, we just stash the
 * information locally.
 */
void
STDC::QMS::Push_Supplier::init (STDC::QMS::CORBA_Handles Handles,
				RtecEventComm::EventType send_event_type,
				RtecEventComm::EventSourceID my_id,
				const char * ec_name_p,
				CORBA_Environment &ACE_TRY_ENV
				)
{
  QMS_Trace ("STDC::QMS::Push_Supplier::init", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Supplier object initializing\n");
  ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);

  this->handles          = Handles;
  this->_send_event_type = send_event_type;
  this->_my_id           = my_id;
  this->EventName        = CORBA::string_dup("Send QMS Message");
  this->Is_Scheduled     = false;
  this->Is_Subscribed    = false;
  this->ec_name_p        = CORBA::string_dup (ec_name_p);

  ACE_TRY
    {
      this->rt_info = 
	this->handles.get_scheduler_handle()->create (this->EventName,
						      ACE_TRY_ENV);
      ACE_TRY_CHECK;
      DEBUG0 (4, "DEBUG: Create a supplier rt info\n"); 
      this->default_schedule();
      this->qos.insert (this->_my_id,	          // source id
			this->_send_event_type, // event type
			this->rt_info,          // handle to the rt info
			1);	// number of calls???
      this->Is_Subscribed    = true;
    }
  ACE_CATCH (CORBA::SystemException, sysex)
    {
      ACE_PRINT_EXCEPTION (sysex, "System Exception supplier::init initializing supplier:");
      this->disconnect();
      ACE_RE_THROW;
    }
  ACE_CATCHANY
    {
      ACE_PRINT_EXCEPTION (ACE_ANY_EXCEPTION, 
			   "Caught exception supplier::init initializing supplier:");
      this->disconnect();
      ACE_RE_THROW;
    }
  ACE_ENDTRY;
}



/**
 *  @brief send event
 *  @param data string to be sent
 *  @return int
 *
 * This is where the payload is sent across packaged in an event. We
 *   - create an event set for one event, initialize event header.
 *   - create the event
 *   - send it across
 */
int
STDC::QMS::Push_Supplier::send(const string &data,
			       CORBA_Environment &ACE_TRY_ENV)
{
  QMS_Trace ("STDC::QMS::Push_Supplier::send", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Supplier object sending message\n");

  ACE_TRY
    {
      // Create an event set for one event, initialize event header.
      RtecEventComm::EventSet event (1);
      event.length (1);
      event[0].header.type   = this->_send_event_type;
      event[0].header.source = this->_my_id;
      event[0].header.ttl    = 1;
      DEBUG0 (4, "DEBUG create dataset\n");
      DEBUG1 (4, "DEBUG Sending Event: %d\n",
	      this->_send_event_type);

      DEBUG1 (4, "DEBUG With ID: %d\n",
	      this->_my_id);

      // Set up time stamps in event header.  This is for performance
      // measurements, so this step can be omitted at will.
      //   ACE_hrtime_t t = ACE_OS::gethrtime ();
      //   ORBSVCS_Time::hrtime_to_TimeT (event[0].header.creation_time, t);
      //   event[0].header.ec_recv_time = ORBSVCS_Time::zero;
      //   event[0].header.ec_send_time = ORBSVCS_Time::zero;

      // Set up message block for event data.
      ACE_Message_Block mb (data.size() + 1);
      // This takes care of positioning the wr_ptr automatically
      char * temp_p = (char *) data.data();
      temp_p[data.size()] = 0;

      mb.copy(temp_p, data.size());
      DEBUG0 (4, "DEBUG Copy data\n");

      // We use replace to minimize the copies, this should result
      // in just one memory allocation;
      event[0].data.payload.replace(data.size(),&mb);


      {
	ACE_Read_Guard<ACE_Thread_Mutex> guard(this->asset_guard);
	DEBUG0 (4, "DEBUG pushing event\n");
	this->_consumer_proxy->push (event, ACE_TRY_ENV);
      }
      ACE_TRY_CHECK;

      // mb.release();
      DEBUG0 (4, "DEBUG done sending\n");
    }
  ACE_CATCH (CORBA::SystemException, sys_ex)
    {
      ACE_PRINT_EXCEPTION (sys_ex, "SYS_EX during send");
      ACE_RE_THROW;
    }
  ACE_CATCHANY
    {
      ACE_PRINT_EXCEPTION (ACE_ANY_EXCEPTION, "NON SYS EX during send");
      ACE_RE_THROW;
    }
  ACE_ENDTRY;


  return 0;
}
/**
 * @brief Get scheduling information
 * @return The current scheduling information set
 */
RtecScheduler::handle_t&
STDC::QMS::Push_Supplier::get_schedule(void)
{
  return this->rt_info;
}

/**
 * @brief Set scheduling information
 * @param schedule the scheduling parameters to set for this object
 */
void
STDC::QMS::Push_Supplier::set_schedule(RtecScheduler::handle_t& schedule)
{
  ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);

  this->rt_info = schedule;
  this->Is_Scheduled = true;
}

/**
 * @brief Get subscription information
 * @return The current subscription information set
 */
ACE_SupplierQOS_Factory&
STDC::QMS::Push_Supplier::get_subscription(void)
{
  return this->qos;
}

/**
 * @brief Set subscription information
 * @param qos the new subsciption informati0on for this object
 */
void
STDC::QMS::Push_Supplier::set_subscription(ACE_SupplierQOS_Factory& qos)
{
  ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);

  this->qos = qos;
  this->Is_Subscribed = true;
}

/**
 * @brief Set up a default scheduling policy for(this object.)
 * 
 *  The execution times are set to reasonable values, but actually
 *  they are changed on the real execution, i.e. we lie to the
 *  scheduler to obtain right priorities; but we don't care if the set
 *  is schedulable. Let us say the execution time for this is 5 milliseconds
 */
void
STDC::QMS::Push_Supplier::default_schedule(CORBA_Environment &ACE_TRY_ENV)
{
  QMS_Trace ("STDC::QMS::Push_Supplier::default_schedule", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Setting a default Schedule\n");

  // No need for a guard element here, since this is a provate
  // function called only by protected members


  ACE_TRY
    {
      ACE_Time_Value tv (0, 5000);
      TimeBase::TimeT time;
      ORBSVCS_Time::Time_Value_to_TimeT (time, tv);
      RtecScheduler::Period_t rate = ACE_U64_TO_U32(time);
      
      DEBUG0 (4, "set the scheduler information\n");
	  DEBUG0 (4, "DEBUG: set the scheduler information\n");
      this->handles.get_scheduler_handle()->set 
	(this->rt_info,
	 RtecScheduler::VERY_HIGH_CRITICALITY,
	 0, 0, 0,
	 rate,
	 RtecScheduler::MEDIUM_IMPORTANCE,
	 0,
	 1,
	 RtecScheduler::OPERATION,
	 ACE_TRY_ENV);
      ACE_TRY_CHECK;
      this->Is_Scheduled = true;
    }
  ACE_CATCH (CORBA::SystemException, sys_ex)
    {
      ACE_PRINT_EXCEPTION (sys_ex, "SYS_EX setting the schedule");
      ACE_RE_THROW;
    }
  ACE_CATCHANY
    {
      ACE_PRINT_EXCEPTION (ACE_ANY_EXCEPTION, "NON SYS EX setting the schedule");
      ACE_RE_THROW;
    }
  ACE_ENDTRY;
}


/**
 *  @brief Connect to the event channel as a supplier
 *  @param ACE_TRY_ENV The ACE exception handling var
 *  @exception CORBA::SystemException
 *
 * This is where the bulk of the configuration occurs. This is where
 * the  admin object is created and the event channel connection is
 * set up 
 */
void
STDC::QMS::Push_Supplier::connect(CORBA::Environment &ACE_TRY_ENV)
  ACE_THROW_SPEC ((CORBA::SystemException))
{
  QMS_Trace ("STDC::QMS::Push_Supplier::connect", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Supplier object connecting\n");
  ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);


  ACE_TRY
    {
      if(this->Is_Scheduled == false)
       {
	 this->default_schedule();
       } /* end of if(this->Is_Scheduled == false) */

      DEBUG2 (4, "DEBUG: My id = %d send event type = %d\n",
	      this->_my_id,  this->_send_event_type);

      if (CORBA::is_nil (this->_supplier_admin.in ()))
       {
	 // We need to get a supplier admin object from the event channel
	 // The canonical protocol to connect to the EC
	 RtecEventChannelAdmin::EventChannel_var event_channel_var;
	 this->handles.get_event_channel(this->ec_name_p, event_channel_var,
					 ACE_TRY_ENV);
	 RtecEventChannelAdmin::EventChannel_ptr event_channel = 
	   event_channel_var.in();
	
	 this->_supplier_admin = event_channel->for_suppliers (ACE_TRY_ENV);
	 ACE_TRY_CHECK;
	 int count = 0;
	 while (CORBA::is_nil (this->_supplier_admin.in ()))
	  {
	    ACE_Time_Value sleep_time (QMS_DEFAULT_retry_sleep_interval, 0);
	    ACE_ERROR ((LM_ERROR,"(%P|%t) Unable to get the admin object.\n"));
	    if(count >= 4)
	     {
	       throw std::runtime_error::runtime_error("Unable to get the admin object.");
	     } /* end of if(count >= 4) */
	  
	    ACE_OS::sleep (sleep_time);
	    DEBUG0 (6, "retrying getting the admin object.\n");

	    this->_supplier_admin =
	      this->handles.get_ec_handle()->for_suppliers (ACE_TRY_ENV);
	    ACE_TRY_CHECK;

	    count++;
	  }
	 DEBUG0 (6, "Got the admin object.\n");
       }
      else
       {
	 DEBUG0 (6, "Already had  the admin object.\n");
       }

      // Now we need the consumer proxy that shall recieve the events
      if (CORBA::is_nil (this->_consumer_proxy.in ()))
       {
	 this->_consumer_proxy =
	   this->_supplier_admin->obtain_push_consumer (ACE_TRY_ENV);
	 ACE_TRY_CHECK;
	 int count = 0;
	 while (CORBA::is_nil (this->_consumer_proxy.in ())){

	   ACE_Time_Value sleep_time (QMS_DEFAULT_retry_sleep_interval, 0);
	   ACE_ERROR ((LM_ERROR," Unable to get the consumer proxy.\n"));
	   if(count >= 4)
	    {
	      throw std::runtime_error::runtime_error("Unable to get the consumer proxy");
	    } /* end of if(count >= 4) */
	  
	   ACE_OS::sleep (sleep_time);
	   DEBUG0 (6, "retrying getting consumer proxy\n");
	   this->_consumer_proxy =
	     this->_supplier_admin->obtain_push_consumer (ACE_TRY_ENV);
	   ACE_TRY_CHECK;
	  
	   count++;
	 }
	 DEBUG0 (6, "Got the consumer proxy.\n");
       }
      else
       {
	 DEBUG0 (6, "Already had  the consumer proxy.\n");
       }

      if (this->connected_to_proxy == false){
	// set ourselves up as suppliers
	RtecEventComm::PushSupplier_var supplier = this->_this (ACE_TRY_ENV);
	ACE_TRY_CHECK;

	DEBUG0 (6, "connecting supplier\n");
	this->_consumer_proxy->connect_push_supplier
	  (supplier.in (), this->qos.get_SupplierQOS (), ACE_TRY_ENV);
	ACE_TRY_CHECK;
	this->connected_to_proxy = true;
	DEBUG0 (6, "supplier connected\n");
      }
    }
  ACE_CATCH (CORBA::SystemException, sysex)
    {
      ACE_PRINT_EXCEPTION (sysex, 
			   "System Exception while connecting Push Supplier:");
      ACE_RE_THROW;
    }
  ACE_CATCHANY
    {
      ACE_PRINT_EXCEPTION (ACE_ANY_EXCEPTION, 
			   "Caught exception while connecting Push Supplier:");
      ACE_RE_THROW;
    }
  ACE_ENDTRY;
}



/**
 *  @brief Disconnect from Event Channel
 *  @param ACE_TRY_ENV ACE exceptiona handling variable
 *  @exception CORBA::SystemException
 *
 * A method to have us disconnect. We need to do a number of things to
 * have us die cleanly.
 *   - Check to see if we are still connected
 *   - tell our supplier to stop sending us events
 *   - when the supplier is disabled, we destroy the proxy
 *   - ditto for the admin object we obtained
 */
void
STDC::QMS::Push_Supplier::disconnect (long etype, CORBA::Environment 
				      &ACE_TRY_ENV)
  ACE_THROW_SPEC ((CORBA::SystemException))
{
  QMS_Trace ("STDC::QMS::Push_Supplier::disconnect", __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Supplier object disconnecting\n");
  ACE_Write_Guard<ACE_Thread_Mutex> guard(this->asset_guard);


  if (CORBA::is_nil (this->_consumer_proxy.in ()))
    return;
  this->_disconnect_event_type = etype;

  if (this->connected_to_proxy == true){
    // Tell the event channel we are going off
    this->_consumer_proxy->disconnect_push_consumer (ACE_TRY_ENV);
    ACE_CHECK;
    this->connected_to_proxy = false;
    DEBUG0 (4, "DEBUG:  Supplier disconnected from proxy\n");
  }
  if (! CORBA::is_nil (this->_consumer_proxy.in ())){
    // This is no longet a valid reference
    this->_consumer_proxy =
      RtecEventChannelAdmin::ProxyPushConsumer::_nil ();
  }

  if (! CORBA::is_nil (this->_supplier_admin.in ())){
    this->_supplier_admin =
      RtecEventChannelAdmin::SupplierAdmin::_nil ();
  }

  // Deactivate the servant
  DEBUG0 (4, "DEBUG:  Supplier deactivating the servant\n");
  PortableServer::ObjectId_var id =
    this->handles.get_root_poa_handle()->servant_to_id (this, ACE_TRY_ENV);
  ACE_CHECK;
  this->handles.get_root_poa_handle()->deactivate_object (id.in (), 
							  ACE_TRY_ENV);
  ACE_CHECK;
}


/**
 * @brief Tell the event channel we are going off the air
 * @param ACE_TRY_ENV ACE exceptiona handling variable
 * @exception CORBA::SystemException
 */
void
STDC::QMS::Push_Supplier::disconnect_push_supplier (CORBA::Environment 
						    &ACE_TRY_ENV)
  ACE_THROW_SPEC ((CORBA::SystemException))
{
  QMS_Trace ("STDC::QMS::Push_Supplier::disconnect_push_supplier",
	     __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Supplier disconnect_push_supplier\n");
  ACE_Read_Guard<ACE_Thread_Mutex> guard(this->asset_guard);

  this->send_disconnect_event(ACE_TRY_ENV);
}


/**
 *  @brief Set the disconnect event to the consumer
 */
void
STDC::QMS::Push_Supplier::send_disconnect_event(CORBA::Environment &ACE_TRY_ENV)
{
  QMS_Trace ("STDC::QMS::Push_Supplier::send_disconnect_event",
	     __LINE__, __FILE__);
  DEBUG0 (4, "DEBUG:  Supplier sending shutdown\n");
  // Send one event shutdown

  ACE_TRY
    {
      RtecEventComm::EventSet event (1);
      event.length (1);
      event[0].header.source = this->_my_id;
      event[0].header.type = this->_disconnect_event_type;


      {
	ACE_Read_Guard<ACE_Thread_Mutex> guard(this->asset_guard);
      	DEBUG0 (4, "DEBUG:    Supplier pushing disconnect\n");
	DEBUG2 (4, "DEBUG: My ID = %d Disconnect event type = %d\n",
		this->_my_id,  this->_disconnect_event_type);

	this->_consumer_proxy->push(event, ACE_TRY_ENV);
	ACE_TRY_CHECK;
      }
    }
  ACE_CATCH (CORBA::SystemException, sys_ex)
    {
      ACE_PRINT_EXCEPTION (sys_ex, "SYS_EX sending the disconnect event");
      ACE_RE_THROW;
    }
  ACE_CATCHANY
    {
      ACE_PRINT_EXCEPTION (ACE_ANY_EXCEPTION, "NON SYS EX sending the disconnect event");
      ACE_RE_THROW;
    }
  ACE_ENDTRY;

  return;
}

/**
 * @brief Return a pointer to the internal corba handles
 * @return A pointer to the internal corba handles
 */
STDC::QMS::CORBA_Handles *
STDC::QMS::Push_Supplier::get_handles(void)
{
  return  & this->handles;
}








